Support --fast-start for Android applications (as an opt-in)

jonahwilliams opened the pull request:
## Description

**TLDR**: Reduces the time it takes to get from `flutter run` to an interactive debugging session. Locally this reduces the startup time for a dirty application from ~40 seconds in the flutter gallery to about ~12 seconds.

The default behavior of flutter run is to build the platform binary (APK, IPA) as necessary and update it on the device before starting a debug session. This artifact is almost always invalidated whenever the user restarts a debugging session, because it depends on all Dart sources, assets, and native dependencies. Depending on the speed of the user's workstation, this can take anywhere from several seconds to several minutes to rebuild.

Both Dart code and Flutter assets can be loaded at runtime from the devfs instead of from the APK - this is how hot restart works. If the Dart source code and assets are removed as inputs from the APK, then invalidations will only occur on Flutter version updates or plugin changes, which will be significantly less frequent than source code changes.

Instead of building the user's application in the initial APK, we build the empty "splash" application which looks like so:

<img src="https://user-images.githubusercontent.com/8975114/69466318-95285700-0d38-11ea-95e2-e8a678240b44.png" width="250">

After syncing files, the tool performs a silent hot restart to begin the debugging session.


~~Known issues: IDEs are not supported because we cannot hot restart into a paused state https://github.com/flutter/flutter/issues/45424~~

Commit:
example of fast start functionality

Commit:
add splash application

Commit:
skip splash build

Commit:
Merge branch 'master' of github.com:flutter/flutter into _fast_start

Commit:
fixes for nulls

Commit:
Update test.dart

feinstein commented:
I think the user should also have the option to --fast-start from a previous version of his own app. this way he can see his own splash screen if he desires so, or any other custom Java code he has.

jonahwilliams commented:
> or any other custom Java code he has.

All of your custom Java code + plugins is bundled in the APK. If you only want to reuse an existing apk, you can use the `--use-application-binary=path/to/app.apk` option.

feinstein commented:
It would be great if all those option can be easily accessible in the IDE.

jonahwilliams commented:
I agree, if it works well --fast-start will likely become the default. I think you should be able to provide application-binary through command line, but I'm not sure how the IDEs expose it (fyi @DanTup  @devoncarew )

feinstein commented:
Yeah, but as far as I understand, this won't be nice to people with custom APKs, which is every one with a splash screen at least.

jonahwilliams commented:
>  this won't be nice to people with custom APKs, which is every one with a splash screen at least.

This builds your APK as is, but with a default flutter application (dart code/assets).

feinstein commented:
Oh, if the Java and etc is left untouched then I don't see why make it different though.

As a curiosity, wouldn't packing the Flutter code in the APK at the first build be quicker than building a default code and then syncing it? Or is it neglectable difference?

jonahwilliams commented:
Consider the following scenario

A user does a clean build installing the full APK on a device and begins a hot reload development cycle. The make several changes to parts of their code, and add a new asset. Finally, they finish development for the day and quit.

The next day, the same user runs `flutter run` again. Because the contents of the application have changed due to their debugging session on the previous day, it needs to be rebuilt and reinstalled. 

With the --fast-start, the same APK is reused because it doesn't depend on the dart code or the assets that were changed, as these are hot restarted into the existing APK. This is faster because it skips rebuild and installing, while the syncing of assets and initial compile needs to be done anyway to start a hot reload debug session.

feinstein commented:
Yes, but I meant with a Hot Restart at each beginning, no need for reinstalls, I believe the hot restart will be faster, as part of the code is already there... Or doesn't it even matter? I don't know if hot restart skips unchanged code, or if it pushes everything anyway.

jonahwilliams commented:
It does sync everything, but when we connect to a device for the first time for debugging, we already have to sync everything anyway

feinstein commented:
I see, so it won't make any difference, thanks @jonahwilliams.

Commit:
add validation tests to fast_start

Commit:
Merge branch '_fast_start' of github.com:jonahwilliams/flutter into _fast_start

Commit:
Merge branch 'master' into _fast_start

Commit:
fix bad merge

Commit:
Update run.dart

jonahwilliams commented:
This merge got out of hand, will fix

Commit:
fix trailing }  again

Commit:
Update run.dart

Commit:
fix test and add error logging

Commit:
Merge branch '_fast_start' of github.com:jonahwilliams/flutter into _fast_start

DanTup commented:
> I think you should be able to provide application-binary through command line, but I'm not sure how the IDEs expose it (fyi @DanTup @devoncarew )

For VS Code, you can pass any additional args through in the launch configuration:

<img width="508" alt="Screenshot 2019-11-27 at 10 01 16 am" src="https://user-images.githubusercontent.com/1078012/69713682-df288880-10fc-11ea-9fa6-927f704da7a4.png">

For options you'd likely want to set globally, we can add settings that let you toggle them on or off without needing a `launch.json` for every project. We'd probably do this for this `--fast-start` so you can easily turn it on for all your projects if you wish (and if it comes the default, you could easily disable it for them all).

DanTup commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -187,6 +187,15 @@ class RunCommand extends RunCommandBase {
+        hide: true,
+        help: 'Whether to quickly bootstrap applications with a minimal app. '
+              'Currently this is only support on Android devices run from the '
+              'command line. This option cannot be pared with --start-paused or '
```
typo? paired?

Also - is this a temporary restriction? IDEs need to use `--start-paused` when debugging.

jonahwilliams replied:
Because of https://github.com/flutter/flutter/issues/45424 we can't yet hot restart an isolate into a paused state (or maybe we can, and it just need a different configuration?). Until that is resolved I can't support IDEs, but it should be fixable...

DanTup commented on packages/flutter_tools/lib/src/run_hot.dart:
```diff
@@ -527,9 +535,10 @@ class HotRunner extends ResidentRunner {
     String reason,
     bool benchmarkMode = false,
+    bool silent = false,
+    bool pause = false, // this is not currently supported for hot restart.
```
Should this comment say "fast start" and not "hot restart"?

jonahwilliams replied:
Its not supported for hot restart confusingly, only hot reload. I've updated the parameter name to better match the usecase

DanTup replied:
Now I'm confused :) Do you just mean when using `--fast-start`? The [daemon docs](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/doc/daemon.md#apprestart) specifically call this out for hot restart:

> `pause`: optional; when doing a hot restart the isolate should enter a paused mode

(though tbh, I'm sure in VS Code we assume that it works fine for both reload and restart... if it doesn't I'm surprised we're not seeing lots of weird issues since it's when we get pause events we re-send the breakpoints)

jonahwilliams replied:
I haven't noticed that working ever ... let me get back to you on that

DanTup replied:
I had a quick test, and it does seem to pause. Here I sent a hot restart with `pause:true` and the isolate went `IsolateRunnable`->`PauseStart`, then I sent breakpoints and `resume`'d and it continued:

<img width="1520" alt="Screenshot 2019-11-28 at 9 19 15 am" src="https://user-images.githubusercontent.com/1078012/69793486-71de2b80-11c0-11ea-9e24-e4fe451da2e3.png">

jonahwilliams replied:
This is great news! Just need to figure out what I was doing wrong...  Thanks @DanTup !

jonahwilliams replied:
TLDR: --start-paused, applies to all isolates, not just the first spawned.

DanTup replied:
Ah yeah, in the IDEs we automatically resume any isolates that appear in the `PauseStart` state as we get events about them (even well after the app has started).

Commit:
Update run.dart

zanderso commented:
Overall looks good. Needs more tests.

Commit:
add integration test for gradle, unit test for runner

Commit:
Merge branch 'master' of github.com:flutter/flutter into _fast_start

Commit:
use tearoffs

jonahwilliams commented on dev/bots/test.dart:
```diff
@@ -284,6 +289,9 @@ Future<void> _runBuildTests() async {
     if (fileEntity is! Directory) {
       continue;
     }
+    if (_excludedExampleApplications.any((String name) => fileEntity.path.contains(name))) {
```
Done

Commit:
Update run.dart

Commit:
Update gradle_fast_start_test.dart

Commit:
Update main.dart

Commit:
Update splash_test.dart

Commit:
Update gradle_fast_start_test.dart

Commit:
Update main.dart

Commit:
Update splash_test.dart

jonahwilliams commented:
Added APK test that verifies that flutter assets are not included, and resident_runner test that verifies fast-start does a hot restart and attaches successfully.

Removed IDE restriction since --start-paused works as expected

DanTup commented on dev/bots/test.dart:
```diff
@@ -754,6 +762,7 @@ Future<void> _runHostOnlyDeviceLabTests() async {
     if (Platform.isLinux) () => _runDevicelabTest('flutter_create_offline_test_linux'),
     if (Platform.isWindows) () => _runDevicelabTest('flutter_create_offline_test_windows'),
     // TODO(ianh): Fails on macOS looking for "dexdump", https://github.com/flutter/flutter/issues/42494
+    () => _runDevicelabTest('gradle_fast_start_test', environment: gradleEnvironment),
```
Is the TODO above this now on the wrong line?

Commit:
Update test.dart

zanderso commented on dev/bots/test.dart:
```diff
@@ -273,6 +273,11 @@ Future<void> _runToolTests() async {
 
+// Example apps that should not be built by _runBuildTests`
+const List<String> _excludedExampleApplications = <String>[
+  'splash',
```
Could you test it without --fast-start?

Could you add a comment explaining the reason to skip it?

jonahwilliams replied:
Added a comment explaining the skip: there is nothing to build for this app

zanderso commented on packages/flutter_tools/lib/src/commands/run.dart:
```diff
@@ -187,6 +187,14 @@ class RunCommand extends RunCommandBase {
+        defaultsTo: false,
+        hide: true,
+        help: 'Whether to quickly bootstrap applications with a minimal app. '
+              'Currently this is only supported on Android devices. This option '
```
Should we do a `throwToolExit` if this is passed for a non-Android device?

jonahwilliams replied:
It's not technically harmful - the app will perform an unnecessary hot restart but otherwise function fine

jonahwilliams replied:
We could either exit, exit + warn, or ignore it

jonahwilliams replied:
I'm thinking we could warn and force it to off. I'd like to avoid requiring the IDE plugins to implement logic for determining if the device supports it, and let the tool fallback to the old behavior automatically. That will make it easier to roll out support for new devices

zanderso replied:
> I'm thinking we could warn and force it to off.

sgtm

jonahwilliams replied:
Added code to force it off, printStatus warning for non machine modes, and test case

zanderso commented on packages/flutter_tools/lib/src/run_hot.dart:
```diff
@@ -211,6 +211,15 @@ class HotRunner extends ResidentRunner {
       }
     }
 
+    if (debuggingOptions.fastStart) {
```
Should this check that the device supports fastStart if we haven't thrown a tool exit?

jonahwilliams replied:
Addressed above

Commit:
address review comments

zanderso reviewed (approved):
lgtm

Commit:
Merge branch 'master' of github.com:flutter/flutter into _fast_start

Commit:
Update run_test.dart

jonahwilliams commented:
The deploy_gallery-macos task is continually failing to start. I'm going to ignore that for this PR, since the code does not effect iOS builds (and tests confirm that)

jonahwilliams commented:
Change was reverted due to test failures, attempting to reland

jonahwilliams commented:
Reland is: https://github.com/flutter/flutter/pull/46140
